improver/utilities/base_classes.py

```diff
@@ -0,0 +1,235 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown copyright. The Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing plugin base classes."""
+
+from abc import ABC, abstractmethod
+from typing import List
+
+from iris.cube import Cube, CubeList
+
+import improver.utilities.cube_descriptor
+from improver.utilities.cube_checker import spatial_coords_match
+
+
+class BasePlugin(ABC):
+    """An abstract class for IMPROVER plugins.
+    Subclasses must be callable. We preserve the process
+    method by redirecting to __call__.
+    """
+
+    def __call__(self, *args, **kwargs):
+        """Makes subclasses callable to use process
+        Args:
+            *args:
+                Positional arguments.
+            **kwargs:
+                Keyword arguments.
+        Returns:
+            Output of self.process()
+        """
+        return self.process(*args, **kwargs)
+
+    @abstractmethod
+    def process(self, *args, **kwargs):
+        """Abstract class for rest to implement."""
+        pass
+
+
+class InputCubesPlugin(BasePlugin):
+    """
+    An abstract class for IMPROVER plugins that generate a new diagnostic.
```

Comment:
My intention is that this plugin can allow us to make all plugins and CLIs agnostic of the input order of cubes. I've added tests showing that it can be combined with the existing `PostProcessingPlugin`.
Your point about intentional non matching is valid. I had already arranged the class so that `assert_time_coords_ok` could be overridden in child classes allowing time-offsets to be checked for. I have now rearranged the class slightly so that `assert_spatial_coords_ok` exists and can be overridden. Both have a bit more info in the doc-string. I've also added an extra boolean `spatial_match` to the `CubeDescriptor` which is used to filter out cubes where spatial matching is inappropriate.
There are now tests for providing one cube descriptor and one cube too, which might allow this plugin to replace `BasePlugin`.